Port realtime servo tutorial

AdamPettinger opened the pull request:
### Description

From issue #39, this ports "Realtime Arm Servoing" for MoveIt2. 

This leans heavily on the demos from [moveit_servo/doc/](https://github.com/ros-planning/moveit2/tree/main/moveit_ros/moveit_servo/doc) which deviate significantly from the ROS1 tutorials.

I also added a way to test servoing with the keyboard, based off the turtlesim teleop demo over in ROS tutorials.

### Open questions from me

1. Most of the code lives in "demos" in the `moveit_servo` code, should those demos be brought to this repo instead?
2. Similar to above, should the `moveit_servo/doc/` directory be removed from the core code there? If I recall, I had originally put it there because the moveit2 tutorials were not off the ground yet
3. How much of the background or "Servo Overview" stuff should actually be included?


### Checklist
- [ ] **Required by CI**: Code is auto formatted using [clang-format](http://moveit.ros.org/documentation/contributing/code)
- [ ] While waiting for someone to review your request, please consider reviewing [another open pull request](https://github.com/ros-planning/moveit2/pulls) to support the maintainers

[//]: # "You can expect a response from a maintainer within 7 days. If you haven't heard anything by then, feel free to ping the thread. Thank you!"

vatanaksoytezer reviewed (commented):
Hi Adam,

Great PR! I didn't test this yet, but will test and leave some other comments as well. For starters here are my opinions on your discussion points:

1. I think demos should be deleted / moved to tutorials. But I suggest we discuss this at today's standup with other folks.
2. I think `moveit_servo/doc` should definitely moved to here.
3. The more the better in my opinion. As far as I know we don't have much documentation other than this, so this would serve a nice guide on what the servo is and how it should be used.

AdamPettinger commented:
I think the consensus is to move as much of the demos/examples/docs from `moveit2` to here in `moveit2_tutorials`. Unless @AndyZe disagrees, I will work on moving the files over and will open a PR removing them from the other repo.

vatanaksoytezer commented:
> I think the consensus is to move as much of the demos/examples/docs from `moveit2` to here in `moveit2_tutorials`. Unless @AndyZe disagrees, I will work on moving the files over and will open a PR removing them from the other repo.

Yes! Sorry I totally forgot to write here after the meeting.

AdamPettinger commented:
I moved the documentation from the source to tutorials and updated the RST to reflect

I also added the C++ interface code, but decided to modify it. It had been creating a class that inherited from rclcpp::Node and using that through the demo. My understanding is that this is the recommended ROS2 design. However for the tutorial, I think it made the code hard to read/understand so I moved to putting everything in the `main` function and using a few global variables. Please let me know if that was a bad idea and I can revert it

MarqRazz commented on package.xml:
```diff
@@ -35,6 +35,7 @@
   <build_depend>tf2_ros</build_depend>
   <build_depend>tf2_eigen</build_depend>
   <build_depend>tf2_geometry_msgs</build_depend>
+  <build_depend>control_msgs</build_depend>
```
Don't we need `moveit_servo` as a build/run depend here too?

AdamPettinger replied:
Yep! Fixed

MarqRazz commented on doc/realtime_servo/realtime_servo_tutorial.rst:
```diff
@@ -1,134 +1,123 @@
 
-And switch to the desired controller with: ::
+    <video width="700px" nocontrols="true" autoplay="true" loop="true">
+        <source src="../../_static/C++_Interface_Demo.webm" type="video/webm">
```
This video wont play on Chrome for me, but works with Firefox.

AdamPettinger replied:
You saw this when you built locally right? I am unsure how to resolve this, but I did notice that another gif/video in the RViz Quickstart tutorial does not play on Chrome when I build locally, but it does play if accessed through [the webpage](http://moveit2_tutorials.picknik.ai/doc/quickstart_in_rviz/quickstart_in_rviz_tutorial.html#moving-joints-or-in-null-space)

MarqRazz replied:
Yes I only checked on a local build.

MarqRazz commented on doc/realtime_servo/realtime_servo_tutorial.rst:
```diff
@@ -1,134 +1,123 @@
 
-If you do not have a SpaceNav 3D mouse, you can publish example servo commands from the command line with: ::
+    <video width="700px" nocontrols="true" autoplay="true" loop="true">
+        <source src="../../_static/Servo_Teleop_Demo.webm" type="video/webm">
```
This video wont play on Chrome for me, but works with Firefox.

MarqRazz reviewed (commented):
Adam, this looks great thanks for all of your hard work here! I have not had a chance to run the code, but here are a few things I noticed.

AndyZe commented on CMakeLists.txt:
```diff
@@ -46,6 +48,8 @@ set(THIS_PACKAGE_INCLUDE_DEPENDS
   pluginlib
   Eigen3
   Boost
+  control_msgs
+  moveit_servo
```
nit: please alphabetize. Sorry for such a nitpick   ;)

AndyZe commented on doc/realtime_servo/realtime_servo_tutorial.rst:
```diff
@@ -1,134 +1,123 @@
+If you do not have a controller, you can still try the demo using your keyboard. With the demo still running, in a new terminal, run ::
 
-In RViz, grab the red/blue/green "interactive marker" and drag the robot to a non-singular position (not all zero joint angles) that is not close to a joint limit. Click "plan and execute" to move the robot to that pose.
+    ros2 run moveit2_tutorials servo_keyboard_input
```
Thanks for adding the keyboard option! Good idea.

AndyZe commented on doc/realtime_servo/realtime_servo_tutorial.rst:
```diff
@@ -1,134 +1,123 @@
-ROS Signals
------------
-An `rqt_graph` of the servo node is shown below (Enlarge by clicking it). Most of these connections can be ignored. The important ones are:
+MoveIt Servo allows you to stream End Effector (EEF) velocity commands to your manipulator and have it execute them concurrently. This enables teleoperation via a wide range of input schemes, or for other autonomous software to control the robot - in visual servoing or closed loop position control for instance.
```
It's weird that this overview is halfway through the document. Recommend merging it with the first section.

AndyZe commented on package.xml:
```diff
@@ -58,6 +60,8 @@
   <run_depend>tf2_eigen</run_depend>
   <run_depend>tf2_geometry_msgs</run_depend>
   <run_depend>xacro</run_depend>
+  <run_depend>control_msgs</run_depend>
+  <run_depend>moveit_servo</run_depend>
```
Nit: alphabetize. Sorry, again.

Can't you combine all of these `build_depend` and `run_depend` to just `depend`?

AdamPettinger replied:
I tried to lump under depend, and the build failed, saying it did not like the 'depend' tag

MarqRazz replied:
Try setting the [package format](https://www.ros.org/reps/rep-0149.html) to `3` at the top of the file

`<package format="3">`

AdamPettinger replied:
That did the trick, thanks!

AndyZe reviewed (approved):
I ran through the commands and everything worked  :+1:

MarqRazz commented:
@AdamPettinger could you resolve that conflict and I think this will be ready to merge.

MarqRazz commented on package.xml:
```diff
@@ -17,47 +17,37 @@
+  <exec_depend>joint_state_publisher</exec_depend>
+  <exec_depend>joy</exec_depend>
+  <!-- <exec_depend>moveit_commander</exec_depend> -->
+  <exec_depend>moveit_fake_controller_manager</exec_depend>
```
This package needs to be removed or it will fail CI. Vatan just [applied](https://github.com/ros-planning/moveit2_tutorials/pull/90/files#diff-37a67ff78eb7260214b323353263b9af40a2fa98719a1e81937fae0159df87a3) this fix in #90

AdamPettinger replied:
Just rebased and removed that depend
